### DISCUSSION TITLE
Can get_byte_array_region() method have a new variant where I can pass the buff length instead of taking buf.len()?

### DISCUSSION THREAD

**Discussion by ghermaniov** [jni-rs/jni-rs#525] [Q&A] [answered]
Hi,

I am currently porting some JNI code from C++ to RUST in an Android app.
In C++ I was using `env->GetByteArrayRegion()` to copy data into a local buffer array with a pre-allocated fixed size, for performance reasons.  The code looked like this:

`env->GetByteArrayRegion(byte_buffer, 0, bytesCount, local_byte_buffer);`

`byte_buffer` and `bytesCount` are coming from Java code through a JNI function, so I know how many bytes I have.
As you can see, in Android's NDK, this method is implemented with 4 parameters but in RUST JNI, `get_byte_array_region()`  (and all the others similar) have just 3 parameters and internally they are using `buf.len()` as the number of bytes to read.

This is forcing me to create a buffer array inside my JNI function with the size equals to `bytes_count`. I would like to be able to re-use a pre-allocated, static, buffer in each call for performance reasons.

In my C++ implementation I was using the JNI tips from Android developer docs:
https://developer.android.com/training/articles/perf-jni#region-calls

Is there a way for me to achieve this?
Thanks!

**Comment by argv-minus-one** [comment, marked as answer]
The `buf` parameter to `JNIEnv::get_*_array_region` is a slice reference, a data structure consisting of a pointer and length. That's why there's only the one parameter instead of two.

If you have a buffer and you want to copy into only part of it, take a slice of that part, using slice notation. Something like this:

```rust
// I assume you have these variables, or something similar, already.
let env: JNIEnv;
let mut local_byte_buffer: Vec<u8>;
let bytesCount: usize;
let byte_buffer: JByteArray;

// Take a slice of the first `bytesCount` bytes of `local_byte_buffer`.
let local_byte_buffer_slice = &mut local_byte_buffer[..bytesCount];

// Now, do the copying.
env.get_byte_array_region(&byte_buffer, 0, local_byte_buffer_slice)?;
```

This will copy into the first `bytesCount` bytes of `local_byte_buffer`.

The slice doesn't have to be at the beginning of the buffer:

```rust
// Take a slice that's `bytesCount` bytes long, starting at offset 42.
let local_byte_buffer_slice = &mut local_byte_buffer[42..(42 + bytesCount)];

// Now, do the copying.
env.get_byte_array_region(&byte_buffer, 0, local_byte_buffer_slice)?;
```

For more information about slices in general, see *The Rust Programming Language* [chapter about slices](https://doc.rust-lang.org/book/ch04-03-slices.html).

**Reply by ghermaniov** [reply]
Hi, 

I am new to Rust, learning as I go, and I completely missed the slices. I will try this.

Thank you very much!
